File: backend/src/flinq/api/dictionary.py
"""Dictionary lookup API (spec Decision 6)."""

from __future__ import annotations

import uuid
from typing import Annotated, Literal

from fastapi import APIRouter, Depends, HTTPException, Query, Request, status
from sqlalchemy.ext.asyncio import AsyncSession

from flinq.core.db import get_session
from flinq.modules.dictionary.links import render_external_links
from flinq.modules.dictionary.provider import WIKTIONARY_ATTRIBUTION, WiktionaryLocalProvider
from flinq.modules.dictionary.schemas import DictionaryLookupResponse, ExternalLinkOut

router = APIRouter(prefix="/api/dictionary", tags=["dictionary"])

LangCode = Literal["en", "ru", "pt"]


def _require_user(request: Request) -> uuid.UUID:
    user_id = getattr(request.state, "user_id", None)
    if user_id is None:
        raise HTTPException(status.HTTP_401_UNAUTHORIZED)
    return user_id


@router.get("/lookup", response_model=DictionaryLookupResponse)
async def lookup(
    request: Request,
    lang: LangCode,
    target: LangCode,
    text: Annotated[str, Query(min_length=1, max_length=256)],
    session: Annotated[AsyncSession, Depends(get_session)],
) -> DictionaryLookupResponse:
    _require_user(request)
    entries = await WiktionaryLocalProvider(session).lookup(text, lang, target)
    links = [
        ExternalLinkOut(name=link.name, url=link.url)
        for link in render_external_links(text, lang, target)
    ]
    return DictionaryLookupResponse(
        entries=entries, attribution=WIKTIONARY_ATTRIBUTION, external_links=links
    )

File: frontend/src/api/dictionary.ts
import { api } from './client'

export interface DictionarySense {
  sense_index: number
  translation: string
  usage_note: string | null
  examples: { text: string; translation: string | null }[]
}
export interface DictionaryEntry {
  headword: string
  part_of_speech: string | null
  senses: DictionarySense[]
}
export interface DictionaryLookup {
  entries: DictionaryEntry[]
  attribution: { source: string; license: string; url: string }
  external_links: { name: string; url: string }[]
}

export const dictionaryApi = {
  lookup: (lang: string, target: string, text: string) => {
    const q = new URLSearchParams({ lang, target, text })
    return api<DictionaryLookup>(`/api/dictionary/lookup?${q.toString()}`)
  },
}
